File: Dynamic_Programming/problem53/problem53.py
from typing import List


class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        it = iter(nums)
        first = next(it)
        cur = first
        best = first

        for x in it:
            cur = max(x, cur + x)
            best = max(best, cur)

        return best

File: Dynamic_Programming/problem53/problem53.cpp
#include <algorithm>
#include <vector>

using namespace std;

class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        int cur = nums[0];
        int best = nums[0];

        for (int i = 1; i < static_cast<int>(nums.size()); ++i) {
            cur = max(nums[i], cur + nums[i]);
            best = max(best, cur);
        }

        return best;
    }
};
